blogposts/2021/optimizing-a-code-intelligence-commit-graph.md

```diff
@@ -0,0 +1,699 @@
+
+Instead of writing the set of visible uploads per commit, what if we only store the visible uploads for commits that can't trivially recomputed? We've already determined the set of commits that can be easily rematerialized - we can just move the rematerialization from database insertion time to query time.
+
+However, it's not as easy as just throwing out the data we don't need - we still need to encode some additional bookkeeping metadata to enable us to recalculate these values cheaply (without pulling back and traversing the entire commit graph on each query).
+
+We introduced a new table, `lsif_nearest_uploads_links`, which stores a link from each commit that can be trivially recomputed to its nearest ancestor with LSIF data. Queries are still now a simply constant-time lookup:
+
+- If the commit exists in `lsif_nearest_uploads`, we simply use those visible uploads, otherwise
+- If the commit exists in `lsif_nearest_upload_links`, we use the visible uploads attached to the ancestor.
+
+#### Example
+
+We'll use the following commit graph again for our example. Here, commit `68acd3` defines upload #1, and `67e0bf` defines upload #2 (both with distinct indexing root directories).
+
+<div class="no-shadow">
+  <img src="https://sourcegraphstatic.com/blog/commit-graph-optimizations/graph5.png" alt="sample commit graph">
+</div>
+
+The `lsif_nearest_uploads` table associates a commit with its visible uploads, just as before. But now, the number of records in the table is much, much smaller. The commits present in this table satisfy one of the properties described above that make the commit non-trivial to recompute.
+
+| repository                    | commit              | upload_ids          |
+| ----------------------------- | ------------------- | ------------------- |
+| github.com/sourcegraph/sample | <code>4a8a33</code> | <code>[]</code>     |
+| github.com/sourcegraph/sample | <code>68acd3</code> | <code>[1]</code>    |
+| github.com/sourcegraph/sample | <code>e43f5b</code> | <code>[1]</code>    |
+| github.com/sourcegraph/sample | <code>67e0bf</code> | <code>[2]</code>    |
+| github.com/sourcegraph/sample | <code>599611</code> | <code>[1, 2]</code> |
+
+Luckily, some benefits compound one another here, and after we started ignoring traversing the graph in both directions, we can simplify these properties to only account for ancestor-direction traversals. Notably, a commit whose parent has multiple children (`7e0471`, for example) were necessary to store for the descendant-direction traversal, but no longer need to be stored. This further increases the number of trivially recomputable commits, saving even more space.
```

Comment:
this paragraph is not clear. what doesn't need to be stored anymore. the multiple children ?